sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/executeImmediate.scala

```diff
@@ -0,0 +1,198 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.analysis
+
+import scala.util.{Either, Left, Right}
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.expressions.{Alias, Expression, VariableReference}
+import org.apache.spark.sql.catalyst.parser.ParseException
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, SetVariable}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern.{EXECUTE_IMMEDIATE, TreePattern}
+import org.apache.spark.sql.connector.catalog.CatalogManager
+import org.apache.spark.sql.errors.QueryCompilationErrors.{invalidNameParameterizedQueryParametersMustBeNamed, unresolvedVariableError, unsupportedParameterExpression}
+import org.apache.spark.sql.types.StringType
+
+/**
+ * Logical plan representing execute immediate query.
+ *
+ * @param args
+ *   parameters of query
+ * @param query
+ *   query string or variable
+ * @param targetVariables
+ *   variables to store the result of the query
+ */
+case class ExecuteImmediateQuery(
+    args: Seq[Expression],
+    query: Either[String, UnresolvedAttribute],
+    targetVariables: Seq[UnresolvedAttribute])
+  extends UnresolvedLeafNode {
+  final override val nodePatterns: Seq[TreePattern] = Seq(EXECUTE_IMMEDIATE)
+}
+
+/**
+ * This rule substitutes execute immediate query node with plan that is passed as string literal
+ * or session parameter.
+ */
+class SubstituteExecuteImmediate(val catalogManager: CatalogManager)
+  extends Rule[LogicalPlan]
+  with ColumnResolutionHelper {
+
+  def resolveVariable(e: Expression): Expression = {
+
+    /**
+     * We know that the expression is either UnresolvedAttribute or Alias, as passed from the
+     * parser. If it is an UnresolvedAttribute, we look it up in the catalog and return it. If it
+     * is an Alias, we resolve the child and return an Alias with the same name.
+     */
+    e match {
+      case u: UnresolvedAttribute =>
+        getVariableReference(u, u.nameParts)
+      case a: Alias =>
+        Alias(resolveVariable(a.child), a.name)()
+      case other =>
+        throw unsupportedParameterExpression(other)
+    }
+  }
+
+  def resolveArguments(expressions: Seq[Expression]): Seq[Expression] = {
+    expressions.map { exp =>
+      if (exp.resolved) {
+        exp
+      } else {
+        resolveVariable(exp)
+      }
+    }
+  }
+
+  def extractQueryString(either: Either[String, UnresolvedAttribute]): String = {
+    either match {
+      case Left(v) => v
+      case Right(u) =>
+        val varReference = getVariableReference(u, u.nameParts)
+
+        if (!varReference.dataType.sameType(StringType)) {
+          throw new AnalysisException(
+            errorClass = "INVALID_VARIABLE_TYPE_FOR_QUERY_EXECUTE_IMMEDIATE",
+            messageParameters = Map("varType" -> varReference.dataType.simpleString))
```

Comment:
Please, quote the data type by `toSQLType` (see other `AnalysisException` in `QueryCompilationErrors`, for instance):
```suggestion
            messageParameters = Map("varType" -> toSQLType(varReference.dataType)))
```